src/WebJobs.Extensions.OpenAI.CosmosDBNoSqlSearch/CosmosDBNoSqlSearchProvider.cs

```diff
@@ -0,0 +1,357 @@
+    }
+
+    /// <summary>
+    /// Add a document to the search index.
+    /// </summary>
+    /// <param name="document">The searchable document.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>Returns a task that completes when the document is successfully saved.</returns>
+    public async Task AddDocumentAsync(
+        SearchableDocument document,
+        CancellationToken cancellationToken
+    )
+    {
+        // Create cosmos client if not exists in the cache.
+        CosmosClient cosmosClient = this.GetCosmosClient(document.ConnectionInfo!.ConnectionName);
+
+        DatabaseResponse databaseResponse = await cosmosClient
+            .CreateDatabaseIfNotExistsAsync(
+                this.cosmosDBNoSqlSearchConfigOptions.Value.DatabaseName,
+                cancellationToken: cancellationToken
+            )
+            .ConfigureAwait(false);
+
+        VectorEmbeddingPolicy vectorEmbeddingPolicy = new VectorEmbeddingPolicy(
+            [
+                new Embedding
+                {
+                    DataType = this.cosmosDBNoSqlSearchConfigOptions.Value.VectorDataType,
+                    Dimensions = this.cosmosDBNoSqlSearchConfigOptions.Value.VectorDimensions,
+                    DistanceFunction = this.cosmosDBNoSqlSearchConfigOptions
+                        .Value
+                        .VectorDistanceFunction,
+                    Path = this.cosmosDBNoSqlSearchConfigOptions.Value.EmbeddingKey,
+                }
+            ]
+        );
+
+        IndexingPolicy indexingPolicy = new IndexingPolicy
+        {
+            VectorIndexes = new Collection<VectorIndexPath>
+            {
+                new()
+                {
+                    Path = this.cosmosDBNoSqlSearchConfigOptions.Value.EmbeddingKey,
+                    Type = this.cosmosDBNoSqlSearchConfigOptions.Value.VectorIndexType,
+                },
+            },
+        };
+        // Create a container if not exists.
+        ContainerProperties containerProperties = new ContainerProperties(
+            document.ConnectionInfo!.CollectionName,
+            "/id"
+        )
+        {
+            VectorEmbeddingPolicy = vectorEmbeddingPolicy,
+            IndexingPolicy = indexingPolicy,
+        };
+
+        ContainerResponse containerResponse = await databaseResponse
+            .Database.CreateContainerIfNotExistsAsync(
+                containerProperties,
+                cancellationToken: cancellationToken
+            )
+            .ConfigureAwait(false);
+
+        await this.UpsertVectorAsync(cosmosClient, document, cancellationToken);
+    }
+
+    CosmosClient GetCosmosClient(string connectionStringName)
+    {
+        CosmosClient cosmosClient;
+        if (
+            !string.IsNullOrEmpty(connectionStringName)
+            && this.cosmosDBClients.TryGetValue(connectionStringName, out cosmosClient)

```

Comment:
this line can be moved inside the if block rather than keeping it empty

Reply:
Moving that will be wrong, because in this if condition we need to check if the connectionStringName is not null and if the dictionary already contains the value retrieve it.

I can update this this with a different approach, where I use containsKey in the if statement and then assign it in the if block.

But normally in multithreaded apps, it's better to use TryGetValue, as it combines both the operation. It avoids the race condition, if the dictionary gets updated between the two separate operations.

Let me know which approach to go with.

Reply:
I wanted to clarify my previous comment regarding line 119. I wasn’t suggesting any changes to lines 117 or 118; my focus was specifically on ensuring that the if block is not left blank.

I agree that using TryGetValue is a good approach, especially in multithreaded contexts. Please make sure to include logic within the if block to handle the value retrieval properly.

Let me know if you have any questions or need further clarification.